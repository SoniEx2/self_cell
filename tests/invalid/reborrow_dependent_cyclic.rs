use std::cell::RefCell;

use self_cell::self_cell;

struct Bar<'a>(RefCell<(Option<&'a Bar<'a>>, String)>);

self_cell! {
    struct Foo {
        owner: (),

        #[not_covariant]
        dependent: Bar,
    }
}

fn main() {
    let mut x = Foo::new((), |_| Bar(RefCell::new((None, "Hello".to_owned()))));

    x.with_dependent(|_, dep| {
        dep.0.borrow_mut().0 = Some(dep);
    });

    x.with_dependent_mut(|_, dep| {
        let r1 = dep.0.get_mut();
        let string_ref_1 = &mut r1.1;
        let mut r2 = r1.0.unwrap().0.borrow_mut();
        let string_ref_2 = &mut r2.1;

        let s = &string_ref_1[..];
        string_ref_2.clear();
        string_ref_2.shrink_to_fit();
        println!("{}", s); // prints garbage
    });
}
