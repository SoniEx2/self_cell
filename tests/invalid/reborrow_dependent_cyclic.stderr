error[E0597]: `x` does not live long enough
  --> $DIR/reborrow_dependent_cyclic.rs:19:5
   |
19 |       x.with_dependent(|_, dep| {
   |       ^ borrowed value does not live long enough
   |  _____|
   | |
20 | |         dep.0.borrow_mut().0 = Some(dep);
21 | |     });
   | |______- argument requires that `x` is borrowed for `'static`
...
34 |   }
   |   - `x` dropped here while still borrowed

error[E0502]: cannot borrow `x` as mutable because it is also borrowed as immutable
  --> $DIR/reborrow_dependent_cyclic.rs:23:5
   |
19 |       x.with_dependent(|_, dep| {
   |       - immutable borrow occurs here
   |  _____|
   | |
20 | |         dep.0.borrow_mut().0 = Some(dep);
21 | |     });
   | |______- argument requires that `x` is borrowed for `'static`
22 |
23 | /     x.with_dependent_mut(|_, dep| {
24 | |         let r1 = dep.0.get_mut();
25 | |         let string_ref_1 = &mut r1.1;
26 | |         let mut r2 = r1.0.unwrap().0.borrow_mut();
...  |
32 | |         println!("{}", s); // prints garbage
33 | |     });
   | |______^ mutable borrow occurs here
